packages/service-client/odsp-client/src/odspClient.ts

```diff
@@ -0,0 +1,205 @@
+	}> {
+		const loader = this.createLoader(containerSchema);
+		const container = await loader.resolve({ url: sharingUrl });
+
+		// eslint-disable-next-line import/no-deprecated
+		const rootDataObject = await requestFluidObject<IRootDataObject>(container, "/");
+		const fluidContainer = new FluidContainer(container, rootDataObject);
+		const services = await this.getContainerServices(container);
+		return { container: fluidContainer, services };
+	}
+
+	private createLoader(containerSchema: ContainerSchema): Loader {
+		const runtimeFactory = new DOProviderContainerRuntimeFactory(containerSchema);
+		const load = async (): Promise<IFluidModuleWithDetails> => {
+			return {
+				module: { fluidExport: runtimeFactory },
+				details: { package: "no-dynamic-package", config: {} },
+			};
+		};
+
+		const codeLoader = { load };
+		const client: IClient = {
+			details: {
+				capabilities: { interactive: true },
+			},
+			permission: [],
+			scopes: [],
+			user: { id: "" },
+			mode: "write",
+		};
+
+		return new Loader({
+			urlResolver: this.urlResolver,
+			documentServiceFactory: this.documentServiceFactory,
+			codeLoader,
+			logger: this.properties.logger,
+			options: { client },
+		});
+	}
+
+	private async createFluidContainer(
+		container: IContainer,
+		connection: OdspConnectionConfig,
+	): Promise<IFluidContainer> {
+		const createNewRequest: IRequest = createOdspCreateContainerRequest(
+			connection.siteUrl,
+			connection.driveId,
+			connection.path,
+			uuid(),
+		);
+
+		// eslint-disable-next-line import/no-deprecated
+		const rootDataObject = await requestFluidObject<IRootDataObject>(container, "/");
+
+		/**
+		 * See {@link FluidContainer.attach}
+		 */
+		const attach = async (): Promise<string> => {
+			if (container.attachState !== AttachState.Detached) {
+				throw new Error("Cannot attach container. Container is not in detached state");
+			}
+			await container.attach(createNewRequest);
+			const resolvedUrl = container.resolvedUrl as IOdspResolvedUrl;
+			if (container.resolvedUrl === undefined) {
+				throw new Error("Resolved Url not available on attached container");
+			}
+			/**
+			 * A unique identifier for the file within the provided RaaS drive ID. When you attach a container,
+			 * a new `itemId` is created in the user's drive, which developers can use for various operations
+			 * like updating, renaming, moving the Fluid file, changing permissions, and more.
+			 */
+			return resolvedUrl.itemId;
+		};
+		const fluidContainer = new FluidContainer(container, rootDataObject);
+		fluidContainer.attach = attach;
+		return fluidContainer;
+	}
+
+	private async getContainerServices(container: IContainer): Promise<OdspContainerServices> {
+		const getAttributes = async (): Promise<OdspServiceAttributes> => {
+			const resolvedUrl = container.resolvedUrl as IOdspResolvedUrl;
+			if (resolvedUrl === undefined) {
+				throw new Error("Resolved Url not available on attached container");
```

Comment:
In general, we always expect these to be defined after attach.  This approach is fine to start with, but it would probably make sense for us to instead await attachment, if the container is not yet attached (rather than reject the promise).

Reply:
adding this to doc feedback section